### PR TITLE
[CALCITE-3568] BigQuery, Hive, Spark SQL dialects do not explicitly support nested aggregates

### PR DESCRIPTION
**Pull request by divyanshuS27**
The BigQuerySqlDialect, HiveSqlDialect, SparkSqlDialect do not support nested aggregates, hence overriding the supportsNestedAggregations() method in them.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 1]
```diff
@@ -108,6 +108,10 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
             && !SqlTypeUtil.isNumeric(call.type);
   }
 
+  @Override public boolean supportsNestedAggregations() {
+    return false;
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: LGTM.

I verified BigQuery by running a query and it does not support nested aggregation. BigQuery doc also does not mention the support of this feature (https://cloud.google.com/bigquery/docs/reference/standard-sql/aggregate_functions)

Comment (zabetak) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Thanks for verifying @amaliujia. One minor thing: for non specific code related things it is better to add comments in the JIRA case. It is better for people who are going to look into the case in the future and it also creates less notifications.

Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Gotcha. Thanks for your suggestion!

File: core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java [hunk 2]
```diff
@@ -70,6 +70,10 @@ public SparkSqlDialect(SqlDialect.Context context) {
     return true;
   }
 
+  @Override public boolean supportsNestedAggregations() {
+    return false;
+  }
+
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java]: IMO, the title of this issue is a little confusing. What I get from the title is BigQuery/Hive/Spark should support nested aggregate, but we don't. However, the fact is BigQuery/Hive/Spark support nested aggregate wrongly. So I suggest changing it. What about `BigQuery, Hive, Spark SQL dialects support nested aggregates wrongly`?

Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java]: +1

Or `BigQuery, Hive, Spark SQL dialects claim they support nested aggregates wrongly`

Comment (divyanshuS27) [reply core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java]: @chunweilei and @amaliujia  point duly noted. Would it be right to say that ### BigQuery, Hive, Spark SQL dialects do not explicitly support nested aggregates?

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 1 above)
Comment (divyanshuS27) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Thank you @amaliujia

File: core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java (hunk 2 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java]: +1 for the new title.

**Comment by chunweilei** [comment]
Could you please resolve the conflicts?

**Comment by divyanshuS27** [comment]
> Could you please resolve the conflicts?

Resolved.

**Comment by chunweilei** [comment]
@divyanshuS27 , there are still conflicts. Please resolve it and also squash commits.

**Comment by amaliujia** [comment]
You can do the following to resolve conflicts and squash commits:
setup Calcite repo you are `upstream`
```
git fetch upstream
git rebase -i upstream/master 
```
After this process, usually I can have a nice and clean commit locally.

**Comment by zabetak** [comment]
I don't think the conflicts are so much of a problem. I would say that before merging this to master somebody has to verify that BigQuery, Hive, Spark SQL indeed do not support nested aggregates.

**Comment by amaliujia** [comment]
I verified Spark SQL does not support nested aggregations
```
spark-sql> SELECT sum(count(key) FROM (select 1 AS key);
Error in query: 
mismatched input 'FROM' expecting {')', ','}(line 1, pos 22)

== SQL ==
SELECT sum(count(key) FROM (select 1 AS key)
----------------------^^^

spark-sql> SELECT sum(count(key)) FROM (select 1 AS key);
Error in query: It is not allowed to use an aggregate function in the argument of another aggregate function. Please use the inner aggregate function in a sub-query.;;
Aggregate [sum(count(key#0)) AS sum(count(key))#3L]
+- SubqueryAlias `__auto_generated_subquery_name`
   +- Project [1 AS key#0]
      +- OneRowRelation
```

**Comment by divyanshuS27** [comment]
Executed  on  Hive.
`hive> select sum(count(1));
FAILED: SemanticException [Error 10128]: Line 1:11 Not yet supported place for UDAF 'count'`

**Pull request closed without merging**